File: abc035_b.kt

fun main(args : Array<String>) {
  val a = readLine()!!.toList()
  var b:MutableList<MutableList<Int>> = mutableListOf(mutableListOf(0, 0))
  a.map { x ->
    when(x) {
      'U' -> { 
        b.map { b2 ->
          b2[1] = b2[1] + 1 
        }
      }
      'D' -> { 
        b.map { b2 ->
          b2[1] = b2[1] - 1 
        }
      }
      'L' -> { 
        b.map { b2 ->
          b2[0] = b2[0] - 1
        }
      }
      'R' -> { 
        b.map { b2 ->
          b2[0] = b2[0] + 1
        }
      }
      '?' -> { 
        val t = b.map { b2 ->
          val b3 = mutableListOf(b2[0], b2[1])
          b3[0] = b3[0] - 1
          val b4 = mutableListOf(b2[0], b2[1])
          b4[0] = b4[0] + 1
          val b5 = mutableListOf(b2[0], b2[1])
          b5[1] = b5[1] + 1
          val b6 = mutableListOf(b2[0], b2[1])
          b6[1] = b6[1] - 1
          listOf(b3, b4, b5, b6)
        }
        b = mutableListOf()
        t.map { t2 -> 
          t2.map { t ->
            b.add(t)
          }
        }
      }
      else -> ""
    }
  }
  val mode = readLine()!!.toInt()
  //println(b)
  val comp = b.map { z ->
    z.map { y -> Math.abs(y)}.reduce { y,x -> y+x }
  }
  when(mode) { 
    1 -> println(comp.max())
    2 -> println(comp.min())
  }
}

File: abc035_b.py

s = input()
T = input()
qs = []
for i, c in enumerate(s):
  if c == '?':
    qs.append( i )

import copy
blocks = [ list(s) ]
for qi in qs:
  _blocks = list()
  for _block in blocks:
    for ch in ['L', 'R', 'U', 'D']:
      cp = copy.copy(_block)
      cp[qi] = ch
      _blocks.append( cp )
    #print(_blocks)
  blocks = _blocks

#print(blocks)

xsc = []
for block in blocks:
  xs = [0,0]
  for ch in block:
    if ch == 'L': xs[0] -= 1
    if ch == 'R': xs[0] += 1
    if ch == 'U': xs[1] += 1
    if ch == 'D': xs[1] -= 1
  #print(xs)
  xsc.append( sum([abs(x) for x in xs]) )

if T == '1':
  print(max(xsc))
else:
  print(min(xsc))
